Scripts/Karazhan: Added chess event

14shagov opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

- Added chess event

**Issues addressed:**

Closes https://github.com/TrinityCore/TrinityCore/issues/7938


**Tests performed:**

build, tested in-game


**Known issues and TODO list:**

- Sometimes several chess pieces fit into the same square
- ~~If a corpse lies for more than 300 seconds, it disappears and will not respawn when the game ends~~

I based it on this: https://github.com/scriptdev2/scriptdev2

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Commit:
chess_event

14shagov committed:
Merge branch '3.3.5' into chess_event

Ovahlord commented on src/server/scripts/EasternKingdoms/Karazhan/karazhan.h:
```diff
@@ -23,40 +23,51 @@
+    DATA_NIGHTBANE             = 11,
+};
+
+enum NAXData
```
nax?

jackpoz commented on src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp:
```diff
@@ -0,0 +1,1353 @@
+{
+    npc_chess_piece_generic(Creature* creature) : ScriptedAI(creature), instance(creature->GetInstanceScript())
+    {
+        Reset();
```
You can calling Reset() in the constructor.

jackpoz commented on src/server/scripts/EasternKingdoms/Karazhan/karazhan.h:
```diff
@@ -23,40 +23,51 @@
 #define KZScriptName "instance_karazhan"
 #define DataHeader "KZ"
 
+#define M_PI_M2       6.28318530718   // 2pi
```
Are you sure there isn't an existing M_PI_M2 already ?

14shagov replied:
I'm not sure if something like this doesn't exist, but I couldn't find it

jackpoz commented:
> I based it on this: https://github.com/scriptdev2/scriptdev2

When you say "I based it on this", how much % would you say it's yours and how much it's the original script ?

aquadeus commented:
actually this event uses a lot of serverside spells that are named and can be scripted to implement it in a more blizzlike way

acidmanifesto commented on src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp:
```diff
@@ -0,0 +1,1353 @@
+        }
+    }
+
+    void AttackStart(Unit* /*victim*/) override { }
```
What is the purpose here?

14shagov replied:
When you right click on an enemy chess piece, your piece just starts walking towards it. This is most likely due to the fact that I'm using possess instead of charm. `void AttackStart(Unit* /*victim*/) override { }` prevents this.

acidmanifesto replied:
Yhe method is complete empty and does nothing.

14shagov replied:
In your opinion, then, what does this method do?
https://github.com/TrinityCore/TrinityCore/blob/c70ceb2ee5a4c9952ec5039b477aead75283c4b2/src/server/game/AI/ScriptedAI/ScriptedCreature.cpp#L146-L152

acidmanifesto commented on src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp:
```diff
@@ -0,0 +1,1353 @@
+
+    void AttackStart(Unit* /*victim*/) override { }
+
+    void EnterEvadeMode(EvadeReason /*why*/) override { }
```
What is the purpose here?

14shagov replied:
Previously, I assumed that the chess piece is constantly in EvadeMode and replenishes its hp because of it. But flag_extra 2 was responsible for this, and now I doubt the necessity of `void EnterEvadeMode(EvadeReason /*why*/) override { }`.

acidmanifesto replied:
The method is completely empty and does nothing.

acidmanifesto commented on src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp:
```diff
@@ -0,0 +1,1353 @@
+                    DoCast(square, SPELL_MOVE_MARKER);
+                    me->GetMotionMaster()->MovePoint(1, square->GetPosition());
+
+                    /*float angle = me->GetRelativeAngle(square);
```
Why is this present?

acidmanifesto commented:
Now with the control AI being vastly different from the scriptdev2 how were you able to test this?

Aokromes commented on sql/updates/world/3.3.5/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,140 @@
+-- Scripts
+UPDATE `creature_template`
+SET `ScriptName` = CASE
```
why no 4 updates?

CraftedRO commented:
> > I based it on this: https://github.com/scriptdev2/scriptdev2
> 
> When you say "I based it on this", how much % would you say it's yours and how much it's the original script ?

I really don't see the relevance of this question, it's absurd to ask him for a percentage in my opinion.

14shagov committed:
fix

jackpoz commented:
> I really don't see the relevance of this question, it's absurd to ask him for a percentage in my opinion.

Possible answers:
- I didn't do anything, it's 100% copied
- I fixed a couple of build errors
- I almost rewrote it from scratch

We want contributions from the original authors and not copypastes found randomly on the internet.

This PR includes quite some programming mistakes and before reviewing all 2267 lines, it would be good to know if any of the review comments will be handled or if they will be a waste of time.

14shagov commented:
> Now with the control AI being vastly different from the scriptdev2 how were you able to test this?

I hope I understood your question correctly. I tested everything on my server.

14shagov commented:
> > I really don't see the relevance of this question, it's absurd to ask him for a percentage in my opinion.
> 
> Possible answers:
> 
> * I didn't do anything, it's 100% copied
> * I fixed a couple of build errors
> * I almost rewrote it from scratch
> 
> We want contributions from the original authors and not copypastes found randomly on the internet.
> 
> This PR includes quite some programming mistakes and before reviewing all 2267 lines, it would be good to know if any of the review comments will be handled or if they will be a waste of time.

I think the second option is more in line with what I did, but it's definitely not a couple build errors. From the scriptdev2 code I took the basic mechanics, like finding the nearest piece or moving a dead piece to the side. I guess to some extent you could call it copypasting.

Here is an example where the code is almost exactly the same:
[scriptdev2](https://github.com/scriptdev2/scriptdev2/blob/4550c59d319a3535eceff9f87ee72fab046577cd/scripts/eastern_kingdoms/karazhan/chess_event.cpp#L400-L438)
[my](https://github.com/14shagov/TrinityCore/blob/003f3a77480a1b8daee328ffda163d970df0f4b3/src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp#L525-L557)

But for example, if you take the mechanics of spells, how they work, who they affect and so on, I wrote it from scratch.

CraftedRO commented:
If "original authors" are dead, that mean TC will never have that code ?

acidmanifesto commented on src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp:
```diff
@@ -0,0 +1,1342 @@
+    SPELL_FULL_HEAL_HORDE           = 39334,
+    SPELL_FULL_HEAL_ALLIANCE        = 39335,
+
+    // Spells used by the chess NPCs    
```
Remove white space at the end as reported by workflow

acidmanifesto commented on src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp:
```diff
@@ -0,0 +1,1342 @@
+    EVENT_MOVE_COMMAND = 1,
+    EVENT_SPELL_COMMAND,
+    EVENT_MOVE,
+};  
```
Remove white space at the end as reported by workflow

Kittnz commented:
> actually this event uses a lot of serverside spells that are named and can be scripted to implement it in a more blizzlike way

Then you can help him out and add those spells.

jackpoz commented:
> If "original authors" are dead, that mean TC will never have that code ?

When this PR will cause bugs and crashes, who will fix those issues ?

If the one sending the PR is not the original author, didn't inspect the code and doesn't understand what's going on, then he will probably not be the one fixing those issues. On the other hand, if the one submitting the PR did go through all the code, understood what the code is doing, spotted and fixed already some issues, then he can be the maintainer of the code included in the PR for the future too.

If the original author is the same as the one submitting the PR, then there is no chance of "I don't know, this is the code I found, I have no clue what it does".

P.S.: This is a generic comment not specifically related to this PR or the author of this PR.

14shagov committed:
fix

14shagov commented:
> > If "original authors" are dead, that mean TC will never have that code ?
> 
> When this PR will cause bugs and crashes, who will fix those issues ?
> 
> If the one sending the PR is not the original author, didn't inspect the code and doesn't understand what's going on, then he will probably not be the one fixing those issues. On the other hand, if the one submitting the PR did go through all the code, understood what the code is doing, spotted and fixed already some issues, then he can be the maintainer of the code included in the PR for the future too.
> 
> If the original author is the same as the one submitting the PR, then there is no chance of "I don't know, this is the code I found, I have no clue what it does".
> 
> P.S.: This is a generic comment not specifically related to this PR or the author of this PR.

I believe I have a good understanding of this code. I have spent a lot of time on it and understand how the event is organized. I can write an instruction for this event inside the `chess_event.cpp` file if needed. I think this will help fix future bugs related to this event in the future.

ckegg commented:
I remember it was scripted by X-Savior

14shagov committed:
guide

Added event guide.
Changed handling of cast primary and secondary chess piece spell casts.
Resolved the problem with corpses.

Aokromes committed:
Merge branch '3.3.5' into chess_event

Aokromes commented:
src/server/scripts/EasternKingdoms/Karazhan/chess_event.cpp:




Multiple blank lines detected, keep only one. Check the files above


/home/circleci/project/src/server/scripts/EasternKingdoms/Karazhan/instance_karazhan.cpp:598:27: fatal error: member access into incomplete type 'Player'
                    player->CastSpell(player, SPELL_GAME_IN_SESSION, true);
                          ^
/home/circleci/project/src/server/game/Scripting/ScriptMgr.h:50:7: note: forward declaration of 'Player'
class Player;
      ^
1 error generated.
make[2]: *** [src/server/scripts/CMakeFiles/scripts_easternkingdoms.dir/build.make:1196: src/server/scripts/CMakeFiles/scripts_easternkingdoms.dir/EasternKingdoms/Karazhan/instance_karazhan.cpp.o] Error 1
